crowdsourcehinter/crowdsourcehinter.py

```diff
@@ -0,0 +1,352 @@
+        """
+        initial = ast.literal_eval(str(data['initial_hints']))
+        generic = ast.literal_eval(str(data['generic_hints']))
+        if type(generic) is list and type(initial) is dict:
+            self.initial_hints = initial
+            self.generic_hints = generic
+            self.Element = str(data['element'])
+            return {'success': True}
+        return {'success': False}
+
+    def resource_string(self, path):
+        """
+        This function is used to get the path of static resources.
+        """
+        data = pkg_resources.resource_string(__name__, path)
+        return data.decode("utf8")
+
+    def get_user_is_staff(self):
+        """
+        Return self.xmodule_runtime.user_is_staff
+        This is not a supported part of the XBlocks API. User data is still
+        being defined. However, It's the only way to get the data right now.
+        """
+        return self.xmodule_runtime.user_is_staff
+
+    def student_view(self, context=None):
+        """
+        This view renders the hint view to the students. The HTML has the hints templated
+        in, and most of the remaining functionality is in the JavaScript.
+        """
+        html = self.resource_string("static/html/crowdsourcehinter.html")
+        frag = Fragment(html)
+        frag.add_javascript_url('//cdnjs.cloudflare.com/ajax/libs/mustache.js/0.8.1/mustache.min.js')
+        frag.add_css(self.resource_string("static/css/crowdsourcehinter.css"))
+        frag.add_javascript(self.resource_string("static/js/src/crowdsourcehinter.js"))
+        frag.initialize_js('CrowdsourceHinter', {'hinting_element': self.Element, 'isStaff': self.xmodule_runtime.user_is_staff})
+        return frag
+
+    @XBlock.json_handler
+    def get_hint(self, data, suffix=''):
+        """
+        Returns hints to students. Hints with the highest rating are shown to students unless the student has already
+        submitted the same incorrect answer previously.
+        Args:
+          data['submittedanswer']: The string of text that the student submits for a problem.
+        returns:
+          'BestHint': the highest rated hint for an incorrect answer
+                        or another random hint for an incorrect answer
+                        or 'Sorry, there are no hints for this answer.' if no hints exist
+          'StudentAnswer': the student's incorrect answer
+        """
+        # populate hint_database with hints from initial_hints if there are no hints in hint_database.
+        # this probably will occur only on the very first run of a unit containing this block.
+        for answers in self.initial_hints:
+            if answers not in self.hint_database:
+                self.hint_database[answers] = {}
+            if self.initial_hints[answers] not in self.hint_database[answers]:
+                self.hint_database[answers].update({self.initial_hints[answers]: 0})
```

Comment:
Minor: this loop would be easier to read if it started with: ```for answers, answer_hints in self.initial_hints.items():```, then you could use ```answer_hints``` instead of ```self.initial_hints[answers]```.